dart/lib/src/transport/rate_limiter.dart

```diff
@@ -0,0 +1,127 @@
+import '../transport/rate_limit_parser.dart';
+
+import '../sentry_options.dart';
+import '../sentry_envelope.dart';
+import '../sentry_envelope_item.dart';
+import '../sentry_item_type.dart';
+import 'rate_limit.dart';
+import 'rate_limit_category.dart';
+
+/// Controls retry limits on different category types sent to Sentry.
+class RateLimiter {
+  RateLimiter(this._clockProvider);
+
+  final ClockProvider _clockProvider;
+  final _rateLimitedUntil = <RateLimitCategory, DateTime>{};
+
+  SentryEnvelope? filter(SentryEnvelope envelope) {
+    // Optimize for/No allocations if no items are under 429
+    List<SentryEnvelopeItem>? dropItems;
+    for (final item in envelope.items) {
+      // using the raw value of the enum to not expose SentryEnvelopeItemType
+      if (_isRetryAfter(item.header.type.toStringValue())) {
+        dropItems ??= [];
+        dropItems.add(item);
+      }
+    }
+
+    if (dropItems != null) {
+      // Need a new envelope
+      final toSend = <SentryEnvelopeItem>[];
+      for (final item in envelope.items) {
+        if (!dropItems.contains(item)) {
+          toSend.add(item);
+        }
+      }
+
+      // no reason to continue
+      if (toSend.isEmpty) {
+        // TODO(denis): Should hints be implemented as well?
+        //markHintWhenSendingFailed(hint, false);
```

Comment:
this is done by the Native SDKs, and Dart SDK does not have offline caching

Reply:
It would be nice though to support offline caching in the Dart SDK for platforms which do not have a native integration.

Reply:
yep, but that would be a different milestone/PR, either natively thru Dart code or our Native SDKs